### PR TITLE
complete table builder

### PR DESCRIPTION
**Pull request by skyzh**
This PR fixes compile errors and some logical errors in current table builder. It is now in sync with current badger master.

I'll demonstrate related changes by commenting on code directly.

The CI failure will be resolved after https://github.com/tikv/agatedb/pull/11/ is ready.

### CODE REVIEW

File: proto/src/proto/meta.proto [hunk 1]
```diff
@@ -55,7 +55,7 @@ message BlockOffset {
 message TableIndex {
   repeated BlockOffset offsets = 1;
   bytes bloom_filter = 2;
-  uint64 estimated_size = 3;
+  uint32 estimated_size = 3;
```
Comment (skyzh) [inline proto/src/proto/meta.proto:58]: Current badger master uses u32.

File: src/format.rs [hunk 2]
```diff
@@ -9,16 +9,10 @@ pub fn key_with_ts(key: impl Into<BytesMut>, ts: u64) -> Bytes {
...
-            buf.as_mut_ptr() as *mut _,
-            8,
-        );
-        key.advance_mut(8);
-    }
+    let mut ts_bytes = key.split_off(key.len());
+    let res = u64::MAX - ts;
+    ts_bytes.put_u64(res);
+    key.unsplit(ts_bytes);
 }
```
Comment (skyzh) [inline src/format.rs:16]: Here I rewrote `append_ts` and `get_ts` in a safe approach. Benchmark result shows that safe version is nearly the same performance as previous unsafe version, and safe `get_ts` is 2x slower than unsafe version. So I retained previous `get_ts` and put safe `append_ts` here.

image

File: src/table.rs [hunk 3]
```diff
@@ -1,5 +1,5 @@
 mod builder;
-
+/*
```
Comment (skyzh) [inline src/table.rs:2]: I'll fix this module later.

File: src/table/builder.rs [hunk 4]
```diff
@@ -114,36 +136,45 @@ impl Builder {
...
+        let mut bytes = BytesMut::new();
+        self.table_index.encode(&mut bytes).unwrap();
         assert!(bytes.len() < u32::MAX as usize);
         self.buf.put_slice(&bytes);
-        self.buf.put_u32_le(bytes.len());
+        self.buf.put_u32_le(bytes.len() as u32);
         let cs = self.build_checksum(&bytes);
         self.write_checksum(cs);
-        self.buf.freeze()
+        self.buf.clone().freeze()
```
Comment (skyzh) [inline src/table/builder.rs:162]: This freeze may introduce overhead. I'll investigate how to fix this later.

File: src/format.rs (hunk 2 above)
Comment (BusyJay) [reply src/format.rs:16]: I think there are problems with criterion or the compiler. I can get an opposite result after moving the legacy case as the first case to bench.

File: src/format.rs [hunk 5]
```diff
@@ -9,16 +9,10 @@ pub fn key_with_ts(key: impl Into<BytesMut>, ts: u64) -> Bytes {
...
-    let buf = key.bytes_mut();
-    unsafe {
-        ptr::copy_nonoverlapping(
-            &res as *const u64 as *const u8,
-            buf.as_mut_ptr() as *mut _,
-            8,
-        );
-        key.advance_mut(8);
-    }
+    let mut ts_bytes = key.split_off(key.len());
```
Comment (BusyJay) [inline src/format.rs:12]: I don't think split is necessary.

File: src/table/builder.rs (hunk 4 above)
Comment (skyzh) [reply src/table/builder.rs:162]: Freezing a `BytesMut` requires full ownership of `buf`. But in `finish(&mut self)`, we could only get a `&mut buf`. I'm not sure whether calling `finish` allows us to give away `builder`. I'll change this to `finish(self)` and remove `clone()` later, if `builder` won't be used after `finish`.

File: Cargo.toml [hunk 6]
```diff
@@ -15,8 +15,15 @@ memmap = "0.7"
...
+[dev-dependencies]
+criterion = "0.3"
+
 [workspace]
 members = [
     "proto",
     "skiplist",
 ]
+
+[[bench]]
```
Comment (BusyJay) [inline Cargo.toml:27]: Should add a new profile for bench. `codegen-units` should be set to 1 to get a correct bench result.

**Review by BusyJay — COMMENTED** [review]
Rest LGTM

File: Cargo.toml (hunk 6 above)
Comment (skyzh) [reply Cargo.toml:27]: I copied TiKV `[profile.bench]` to `Cargo.toml`

**Review by BusyJay — APPROVED** [review]

**Pull request merged**